### PR TITLE
PWGHF: taskFlow.cxx: Add process functions for data TPC-MFT HF-h case, and MC Rec/Gen for TPC-TPC h-h case

### PR DESCRIPTION
**Pull request by AlexianL**
QA functions and histograms for the data TPC-MFT HF-h case were also added
In both TPC-MFT h-h and TPC-MFT HF-h case, histograms have been more clearly separated to investigate tracks/candidates from the TPC and tracks from the MFT
QA functions and histograms for the MC Rec/Gen TPC-TPC h-h case were also added
Also mixCollisions function had to be duplicated to be adapted for MC Gen

### CODE REVIEW

File: PWGHF/HFC/Tasks/taskFlow.cxx [hunk 1]
```diff
@@ -13,6 +13,8 @@
 /// \author Katarina Krizkova Gajdosova <katarina.gajdosova@cern.ch>, CERN
 /// \author Maja Kabus <maja.kabus@cern.ch>, CERN
 
+#include <iostream>
+
```
Comment (vkucera) [inline PWGHF/HFC/Tasks/taskFlow.cxx:17]: Why? This should never be needed in O2.

File: PWGHF/HFC/Tasks/taskFlow.cxx [hunk 2]
```diff
@@ -47,6 +50,8 @@ using namespace o2::framework;
 using namespace o2::framework::expressions;
 
 struct HfTaskFlow {
+  Service<o2::framework::O2DatabasePDG> pdg;
+
```
Comment (vkucera) [inline PWGHF/HFC/Tasks/taskFlow.cxx:54]: Please order the struct members according to the HF conventions.

Comment (vkucera) [reply PWGHF/HFC/Tasks/taskFlow.cxx:54]: Same for the other added members.

File: PWGHF/HFC/Tasks/taskFlow.cxx [hunk 3]
```diff
@@ -271,8 +527,41 @@ struct HfTaskFlow {
...
+  {
+    if constexpr (std::is_same_v<TracksWDcaSelMC, TTracks>) { // if MC Rec
+      registry.fill(HIST("MC/Rec/TpcTpc/HadronHadron/MixedEvent/hMultiplicityMixing"), multiplicity);
+      registry.fill(HIST("MC/Rec/TpcTpc/HadronHadron/MixedEvent/hVtxZMixing"), vz);
+    } else { // if MC Gen
+      registry.fill(HIST("MC/Gen/TpcTpc/HadronHadron/MixedEvent/hMultiplicityMixing"), multiplicity);
+      registry.fill(HIST("MC/Gen/TpcTpc/HadronHadron/MixedEvent/hVtxZMixing"), vz);
+    }
+
+    int Ntracks = 0;
```
Comment (vkucera) [inline PWGHF/HFC/Tasks/taskFlow.cxx:542]:
```suggestion
    int nTracks = 0;
```

File: PWGHF/HFC/Tasks/taskFlow.cxx [hunk 4]
```diff
@@ -271,8 +527,41 @@ struct HfTaskFlow {
...
+    for (const auto& track1 : tracks) {
+      Ntracks++;
```
Comment (vkucera) [inline PWGHF/HFC/Tasks/taskFlow.cxx:544]: Why can't you use `tracks.size()`?

File: PWGHF/HFC/Tasks/taskFlow.cxx [hunk 5]
```diff
@@ -271,8 +527,41 @@ struct HfTaskFlow {
...
+        registry.fill(HIST("MC/Gen/TpcTpc/HadronHadron/MixedEvent/hPtMixing"), track1.pt());
+        registry.fill(HIST("MC/Gen/TpcTpc/HadronHadron/MixedEvent/hEtaMixing"), track1.eta());
+        registry.fill(HIST("MC/Gen/TpcTpc/HadronHadron/MixedEvent/hPhiMixing"), track1.phi());
+      }
+    }
+    if constexpr (std::is_same_v<TracksWDcaSelMC, TTracks>) { // if MC Rec
+      registry.fill(HIST("MC/Rec/TpcTpc/HadronHadron/MixedEvent/hNtracksMixing"), Ntracks);
+    } else { // if MC Gen
+      registry.fill(HIST("MC/Gen/TpcTpc/HadronHadron/MixedEvent/hNtracksMixing"), Ntracks);
+    }
```
Comment (vkucera) [inline PWGHF/HFC/Tasks/taskFlow.cxx:559]: How does this guarantee that the `else` case is MC gen?

File: PWGHF/HFC/Tasks/taskFlow.cxx [hunk 6]
```diff
@@ -312,9 +725,30 @@ struct HfTaskFlow {
...
+  {
+    //  remove MC particles with charge = 0
+    int8_t sign = 0;
+    TParticlePDG* pdgparticle = pdg->GetParticle(track.pdgCode());
+    if (pdgparticle != nullptr) {
+      sign = (pdgparticle->Charge() > 0) ? 1.0 : ((pdgparticle->Charge() < 0) ? -1.0 : 0.0);
+    }
+    if (sign == 0) {
+      return false;
+    }
```
Comment (vkucera) [inline PWGHF/HFC/Tasks/taskFlow.cxx:739]: Why don't you just check directly `pdgparticle->Charge() == 0`?

File: PWGHF/HFC/Tasks/taskFlow.cxx [hunk 7]
```diff
@@ -563,16 +1240,25 @@ struct HfTaskFlow {
       return size;
     };
 
+    /*
+    auto getTracksSize = [&candidates, this](FilteredCollisionsWSelMult::iterator const& col) {
+      auto associatedTracks = candidates.sliceByCached(o2::aod::track::collisionId, col.globalIndex(), this->cache);
+      auto size = associatedTracks.size();
+      return size;
+    };
+    */
```
Comment (vkucera) [inline PWGHF/HFC/Tasks/taskFlow.cxx:1249]: Better to use `//` on each line. It's better for parsing per line with code analysis tools.

**Review by vkucera — CHANGES_REQUESTED** [review]
Hi @AlexianL , please see some first comments.

File: PWGHF/HFC/Tasks/taskFlow.cxx (hunk 5 above)
Comment (AlexianL) [reply PWGHF/HFC/Tasks/taskFlow.cxx:559]: It is within the MC QA function, and only MC Rec uses "TracksWDcaSelMC" as TTracks, while MC Gen uses a different type "aodMcParticles"

File: PWGHF/HFC/Tasks/taskFlow.cxx (hunk 1 above)
Comment (AlexianL) [reply PWGHF/HFC/Tasks/taskFlow.cxx]: Forgot to remove it, done

File: PWGHF/HFC/Tasks/taskFlow.cxx (hunk 2 above)
Comment (AlexianL) [reply PWGHF/HFC/Tasks/taskFlow.cxx]: Changed location of that line

File: PWGHF/HFC/Tasks/taskFlow.cxx (hunk 4 above)
Comment (AlexianL) [reply PWGHF/HFC/Tasks/taskFlow.cxx]: No specific reason to not use it, so I changed the code to use it

File: PWGHF/HFC/Tasks/taskFlow.cxx (hunk 6 above)
Comment (AlexianL) [reply PWGHF/HFC/Tasks/taskFlow.cxx]: No specific reason, so I use directly the charge equal 0 now

File: PWGHF/HFC/Tasks/taskFlow.cxx (hunk 7 above)
Comment (AlexianL) [reply PWGHF/HFC/Tasks/taskFlow.cxx]: Replaced all /**/ with //

**Comment by AlexianL** [comment]
> Error while checking build/O2Physics/o2 for [9804a42](https://github.com/AliceO2Group/O2Physics/commit/9804a4200fa7d69341198c0ff9c781a606aab504) at 2024-09-27 19:17:
> 
> ```
> ## sw/BUILD/O2Physics-latest/log
> /sw/SOURCES/O2Physics/7802-slc7_x86-64/0/PWGHF/HFC/Tasks/taskFlow.cxx:544:7: error: 'Ntracks' was not declared in this scope; did you mean 'nTracks'?
> /sw/SOURCES/O2Physics/7802-slc7_x86-64/0/PWGHF/HFC/Tasks/taskFlow.cxx:556:83: error: 'Ntracks' was not declared in this scope; did you mean 'nTracks'?
> /sw/SOURCES/O2Physics/7802-slc7_x86-64/0/PWGHF/HFC/Tasks/taskFlow.cxx:558:83: error: 'Ntracks' was not declared in this scope; did you mean 'nTracks'?
> ninja: build stopped: subcommand failed.
> ```
> 
> Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/O2Physics/7802/9804a4200fa7d69341198c0ff9c781a606aab504/build_O2Physics_o2/pretty.html).

Not sure to understand the problem here, all "Ntracks" were already changed to "nTracks"

**Review by zchochul — DISMISSED** [review]

**Comment by AlexianL** [comment]
> Error while checking build/O2Physics/o2 for [edd7cf7](https://github.com/AliceO2Group/O2Physics/commit/edd7cf7da0fe859887cbd6959e00ca5d11ad48b8) at 2024-10-04 12:53:
> 
> ```
> ## sw/BUILD/O2Physics-latest/log
> /sw/SOURCES/O2Physics/7802-slc7_x86-64/0/DPG/Tasks/AOTTrack/qaTrackSplitting.cxx:170:128: fatal error: error writing to /tmp/cc2FfA0n.s: No space left on device
> /sw/SOURCES/O2Physics/7802-slc7_x86-64/0/DPG/Tasks/AOTTrack/MonitorFilterBit.cxx:585:1: fatal error: error writing to /tmp/ccoQz47A.s: No space left on device
> /sw/SOURCES/O2Physics/7802-slc7_x86-64/0/DPG/Tasks/AOTTrack/PID/TOF/qaPIDTOFBeta.cxx:418:124: fatal error: error writing to /tmp/cc3MBysw.s: No space left on device
> /sw/SOURCES/O2Physics/7802-slc7_x86-64/0/DPG/Tasks/AOTTrack/PID/HMPID/analysisHMPID.cxx:131:126: fatal error: error writing to /tmp/ccU5sbEt.s: No space left on device
> /sw/SOURCES/O2Physics/7802-slc7_x86-64/0/DPG/Tasks/AOTTrack/PID/Combined/qaPIDWithV0s.cxx:265:1: fatal error: error writing to /tmp/ccHQuxS7.s: No space left on device
> /sw/SOURCES/O2Physics/7802-slc7_x86-64/0/DPG/Tasks/AOTTrack/PID/TPC/qaPIDTPCSignal.cxx:299:126: fatal error: error writing to /tmp/ccKJlwRe.s: No space left on device
> /sw/SOURCES/O2Physics/7802-slc7_x86-64/0/DPG/Tasks/AOTTrack/PID/HMPID/qaHMPID.cxx:305:120: fatal error: error writing to /tmp/ccfUF9QC.s: No space left on device
> /sw/SOURCES/O2Physics/7802-slc7_x86-64/0/DPG/Tasks/AOTTrack/PID/TOF/qaPIDTOFEvTime.cxx:604:133: fatal error: error writing to /tmp/ccT7GX8n.s: No space left on device
> /sw/SOURCES/O2Physics/7802-slc7_x86-64/0/DPG/Tasks/AOTTrack/qaEventTrackLiteProducer.cxx:432:1: fatal error: error writing to /tmp/cctEfcEn.s: No space left on device
> /sw/SOURCES/O2Physics/7802-slc7_x86-64/0/DPG/Tasks/AOTTrack/PID/TOF/qaPIDTOFMC.cxx:752:122: fatal error: error writing to /tmp/ccP4dGA8.s: No space left on device
> /sw/SOURCES/O2Physics/7802-slc7_x86-64/0/DPG/Tasks/AOTTrack/PID/TPC/qaPIDTPCMC.cxx:591:122: fatal error: error writing to /tmp/ccOgNL1E.s: No space left on device
> /sw/SOURCES/O2Physics/7802-slc7_x86-64/0/DPG/Tasks/AOTTrack/qaMatchEff.cxx:3448:1: fatal error: error writing to /tmp/ccGD8w3s.s: No space left on device
> /sw/SOURCES/O2Physics/7802-slc7_x86-64/0/DPG/Tasks/AOTTrack/qaImpPar.cxx:717:1: fatal error: error writing to /tmp/ccZolJZ3.s: No space left on device
> /sw/SOURCES/O2Physics/7802-slc7_x86-64/0/DPG/Tasks/AOTTrack/qaEventTrack.cxx:1733:1: fatal error: error writing to /tmp/ccuYJH34.s: No space left on device
> /sw/SOURCES/O2Physics/7802-slc7_x86-64/0/DPG/Tasks/AOTTrack/qaEfficiency.cxx:2118:124: fatal error: error writing to /tmp/ccMbX10i.s: No space left on device
> ninja: build stopped: subcommand failed.
> ```
> 
> Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/O2Physics/7802/edd7cf7da0fe859887cbd6959e00ca5d11ad48b8/build_O2Physics_o2/pretty.html).

It is not obvious to me what the problem is here

**Comment by vkucera** [comment]
@AlexianL , the machine that checks the compilation ran out of disk space. Unlikely your fault.

**Comment by vkucera** [comment]
@AlexianL , please always compile and test your code before making/updating a PR. Your code does not even compile, wasting resources for PR tests.

File: PWGHF/HFC/Tasks/taskFlow.cxx (hunk 6 above)
Comment (AlexianL) [reply PWGHF/HFC/Tasks/taskFlow.cxx:727]: After further investigation, using pdgparticle->Charge() == 0 directly leads to a segmentation fault, so I came back to the previous version

Comment (vkucera) [reply PWGHF/HFC/Tasks/taskFlow.cxx:727]: That is very unlikely.

File: PWGHF/HFC/Tasks/taskFlow.cxx [hunk 8]
```diff
@@ -63,25 +65,68 @@ struct HfTaskFlow {
...
+  using FilteredCollisionsWSelMultMC = soa::Filtered<soa::Join<aod::Collisions, aod::McCollisionLabels, aod::EvSels, aod::Mults>>;
+  using TracksWDcaSelMC = soa::Filtered<soa::Join<aod::TracksWDca, aod::TrackSelection, aod::McTrackLabels>>;
+
+  // using aodCollisions = soa::Filtered<soa::Join<aod::Collisions, aod::EvSels, aod::Mults, aod::McCollisionLabels>>;
+
+  using aodMcCollisions = soa::Filtered<aod::McCollisions>;
+
+  using aodMcParticles = soa::Filtered<aod::McParticles>;
+
+  using aodTracks = soa::Filtered<soa::Join<aod::Tracks, aod::TracksDCA, aod::TrackSelection, aod::McTrackLabels>>;
```
Comment (vkucera) [inline PWGHF/HFC/Tasks/taskFlow.cxx:114]: These names are too vague, misleading and not in line with the naming conventions.
Use names that clearly indicate what the table contains, e.g. `FilteredMcCollisions`, `FilteredMcParticles`, `FilteredTracksWDcaSelMcLabels`.

Comment (vkucera) [inline PWGHF/HFC/Tasks/taskFlow.cxx:114]: Fix the order of struct members.

File: PWGHF/HFC/Tasks/taskFlow.cxx [hunk 9]
```diff
@@ -100,90 +145,262 @@ struct HfTaskFlow {
...
+    //  =========================
     constexpr int kNBinsEvents = 3;
     registry.add("Data/hEventCounter", "hEventCounter", {HistType::kTH1F, {{kNBinsEvents, 0.5, 0.5 + kNBinsEvents}}});
     //  set axes of the event counter histogram
     std::string labels[kNBinsEvents];
     labels[0] = "all";
     labels[1] = "after trigger selection (Run 2)";
     labels[2] = "after Physics selection";
+
+    const int maxMixBin = axisMultiplicity->size() * 14; // 14 bins for z-vertex
```
Comment (vkucera) [inline PWGHF/HFC/Tasks/taskFlow.cxx:179]: Isn't this just the number of mixing bins? If yes, I would suggest to call it `nBinsMix`.

File: PWGHF/HFC/Tasks/taskFlow.cxx [hunk 10]
```diff
@@ -195,107 +412,287 @@ struct HfTaskFlow {
...
       registry.fill(HIST("Data/TpcTpc/HadronHadron/SameEvent/hEtaPhi"), multiplicity, track1.eta(), track1.phi());
     }
-    registry.fill(HIST("Data/TpcTpc/HadronHadron/SameEvent/hNtracks"), Ntracks);
+    registry.fill(HIST("Data/TpcTpc/HadronHadron/SameEvent/hNtracks"), nTracks);
   }
 
+  // ---- MC : TPC-TPC h-h Same Event QA ----
+  // Changed quickly void for int type + added the return for a test
+  template <CorrelationContainer::CFStep step, typename TTracks>
+  int fillTpcTpcChChSameEventQAmc(float multiplicity, TTracks const& tracks)
```
Comment (vkucera) [inline PWGHF/HFC/Tasks/taskFlow.cxx:474]:
```suggestion
  int fillTpcTpcChChSameEventQaMc(float multiplicity, TTracks const& tracks)
```

File: PWGHF/HFC/Tasks/taskFlow.cxx [hunk 11]
```diff
@@ -508,33 +1069,131 @@ struct HfTaskFlow {
...
+    // BinningPolicyBase<2> baseBinning{{axisVertex, axisMultiplicity}, true};
+    // int bin = baseBinning.getBin(std::make_tuple(mcCollision.posZ(), multiplicity));
+    // registry.fill(HIST("MC/Gen/TpcTpc/HadronHadron/SameEvent/hEventCountSame"), bin);
+
+    //  fill correlations for MC collisions that have a reconstructed collision
+    // got rid of the second const auto for multPrimaryCharge0
+    // This line below for sure induce that some plots are filled two times
+    // multPrimaryCharge0 = fillTpcTpcChChSameEventQAmc<CorrelationContainer::kCFStepVertex>(multiplicity, mcParticles);
+    // sameTPCTPCChChMC->fillEvent(multPrimaryCharge0, CorrelationContainer::kCFStepVertex);
+    // fillCorrelations<CorrelationContainer::kCFStepVertex>(sameTPCTPCChChMC, mcParticles, mcParticles, multPrimaryCharge0, mcCollision.posZ());
```
Comment (vkucera) [inline PWGHF/HFC/Tasks/taskFlow.cxx:1188]: Try to discuss with Katarina which code is used which is not and remove the commented parts.

**Review by vkucera — CHANGES_REQUESTED** [review]

File: PWGHF/HFC/Tasks/taskFlow.cxx (hunk 11 above)
Comment (AlexianL) [reply PWGHF/HFC/Tasks/taskFlow.cxx:1188]: I am currently not in contact with Katarina and I don't really know when I will be able to contact her. Should I simply remove completely the comment until then, or can I leave it so I am sure to not forget about it ?

File: PWGHF/HFC/Tasks/taskFlow.cxx (hunk 10 above)
Comment (AlexianL) [reply PWGHF/HFC/Tasks/taskFlow.cxx:474]: Other QA functions are similarly called "fill.....SameEventQA", should I change to "fill....SameEventQa" for all of them or that recommended change is only for MC functions ?

Comment (vkucera) [reply PWGHF/HFC/Tasks/taskFlow.cxx:474]: If you can change `QA` -> `Qa` in the other names as well, that would be great, thanks.

File: PWGHF/HFC/Tasks/taskFlow.cxx (hunk 11 above)
Comment (vkucera) [reply PWGHF/HFC/Tasks/taskFlow.cxx:1188]: It's OK to keep it commented out until you get in touch with Katarina.

File: PWGHF/HFC/Tasks/taskFlow.cxx (hunk 8 above)
Comment (AlexianL) [reply PWGHF/HFC/Tasks/taskFlow.cxx]: I changed the names according to your proposition

Comment (AlexianL) [reply PWGHF/HFC/Tasks/taskFlow.cxx]: I sorted everything in a clearer way 
- using declarations for both DATA/MC and in alphabetical order
- Filters & partitions for both DATA/MC and in alphabetical order
- preslices for both DATA/MC and in alphabetical order

File: PWGHF/HFC/Tasks/taskFlow.cxx (hunk 9 above)
Comment (AlexianL) [reply PWGHF/HFC/Tasks/taskFlow.cxx]: I think so too, I renamed it nBinsMix

**Review by vkucera — APPROVED** [review]
Thanks @AlexianL for implementing patiently my comments.
I leave the final approval and merging to the HFC coordinators.

**Review by deepathoms — APPROVED** [review]

**Pull request merged**